src/twisted/conch/newsfragments/12046.bugfix

```diff
@@ -0,0 +1 @@
+twisted.conch.insults.window.Widget.functionKeyReceived fix a regression introduced with #8214 in Twisted 16.5.0
```

Comment:
Maybe this summarize better this change

```suggestion
twisted.conch.insults.window.Widget.functionKeyReceived now dispatched functional key events to method with the name `func_KEYNAME`, where `KEYNAME` can be `F1`, `F2`, or `UP_ARROW`. This is a regression introduced with #8214 in Twisted 16.5.0, in which `KEYNAME` was handled with square brackets like `[F1]`.
```

Reply:
TL;DR: I think this begs for deeper refactoring and this PR fixes just the regression

I have the rest of updates prepared, however reading this made me to re-read whole part in `insults` to check how it is done. Now i really think moving from `const` objects to plain `bytes` in Twisted 16.5.0 when handling special keys was really a step back and proper way to fix this would be to reintroduce `const` objects back. It just feels wrong when i have to blindly "unpack" some "constant" from string value.

1. List of function keys "names" (what was once a source for name part of the constant) is considered implementation detail in `insults.insults._KEY_NAMES` and even there it was considered as a bad placement with following comment:

   ```python
   # XXX TODO - These attributes are really part of the
   # ITerminalTransport interface, I think.
   ```
2. My test now "randomly" uses some keys, because it'd be lot of copy&paste to test them all. Even then, you can't be sure if that's all of them, because there is no single source of truth for them.
3. Setting keys as attributes directly on `insults.ServerProtocol` is also strange, counterintuitive and potential source of weird bugs. Should be in it's own namespace and then you could list them easily from anywhere.
4. Same list of key names is used on two other places — `insults.helper` and `conch.recvline`; these places are identical and weird on their own, as bytes are concatenated to form a string that's converted to plain `str` and then put through `exec()`, just to introduce new variable based on the name and having plain `object()` as value. Original `exec()` code dates back 19 years (shouldn't be excuse to use `exec` even then), however useless `bytes` → `str` conversion is far newer. Simple `vars()["F1"] = object()` would be enough (or `locals()` / `globals()`).

Reply:
Aggree. The keyname  part is strange and confusing. Thanks for the info.
With this bug, it looks like since 2016 nobody is using this code.
Otherwise I was expecting to see this bug report sooner.

So I think that it safe to make backward incompatible changes, if you find it easier to use this code.

I am not using this code and don't plan to use it. Just trying to help with a review here.